### PR TITLE
IoT Hub - X509CA support

### PR DESCRIPTION
**Pull request by mtuchkov**
Migrating these changes: [1253](https://github.com/Azure/azure-rest-api-specs/pull/1253)
to a new API version.

This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 

### PR information
- [ ] The title of the PR is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [ ] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes.
- [ ] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [ ] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [ ] My spec meets the review criteria:
  - [ ] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [ ] The spec follows the guidelines described in the [Swagger checklist](../documentation/swagger-checklist.md) document.
  - [ ] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

### CODE REVIEW

**Comment by mtuchkov** [comment]
@anuchandy The error that linter shows are applicable to the first class ARM resources. LIke, 'Top level properties should be one of name, type, id, location, properties, tags, plan, sku, etag, managedBy, identity'. The certificates are iothub's subordinate reource. The warnings are not applicable also, for instance on PUT we accept the Certificate as a PEM or CER files, but on GET we give back a normalized JSON representation of the certificate with Subject, Expiry, etc.

Moreover, these PR was already approved two months back. See [1253](https://github.com/Azure/azure-rest-api-specs/pull/1253) Pull Request

**Comment by anuchandy** [comment]
@mtuchkov my understanding is this operations group which represents operations on the certificates is already deployed but it is not exposed through swagger, so no SDK exposes these functionalities.

In order to comply with the new process we have, I would start with feed back from ARM and their sign off. Added the tag "WaitForARMFeedBack".

**Comment by mtuchkov** [comment]
Hi @anuchandy,
@andrew-buckley is working on SDK for this feature and it is already done and is waiting for swagger to be merged in to publish the SDK.

**Comment by anuchandy** [comment]
@mtuchkov thanks.

@ravbhatnagar could you please sign-off from ARM side?, this is the first time operation group `Certificates` is going to be exposed (though it is already deployed) via SDKs. 'Certificates` is child resource of ARM resource `IotHub`

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (anuchandy) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: It is recommended to use plural for operation group name, so this becomes `Certificates_Get`

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (anuchandy) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: Please fix indentation

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (anuchandy) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: Is this a long running operation?

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (anuchandy) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: Is delete a long running operation?

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (anuchandy) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: For better user experience this operation Id should be `Certificates_ListByIoTHub`, this will ensure all the operations on certificates grouped under the `Certificates` class/interface.

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (anuchandy) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: Curious does the service return `DefaultErrorResponse` in response to a List request? if so under what cases it can happen

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (anuchandy) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: Certificates_CreateOrUpdate

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (anuchandy) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: Certificates_Delete

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (anuchandy) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: Certificates_GenerateVerificationCode

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (anuchandy) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: Is this a Long running operation, please confirm..

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (anuchandy) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: Does service returns 204 if resource not exists? if yes please add it as a response entry .

**Review by anuchandy — CHANGES_REQUESTED** [review]
Can you please provide examples for the operations on certificates? see https://github.com/Azure/azure-rest-api-specs/tree/current/specification/compute/resource-manager/Microsoft.Compute/2017-03-30/examples

**Comment by anuchandy** [comment]
@mtuchkov looks like there is another PR opened with similar changes https://github.com/Azure/azure-rest-api-specs/pull/1487.. 

Could you take a look at linter error and warnings.. https://travis-ci.org/Azure/azure-rest-api-specs/jobs/261929893

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (anuchandy) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: Please use value of `tags` collection in `all operations` as operation group name i.e. `Certitifcates`

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (mtuchkov) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: Any kind of error, including 500.

---
In reply to: [132583205](https://github.com/Azure/azure-rest-api-specs/pull/1514#discussion_r132583205) [](ancestors = 132583205)

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (mtuchkov) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: Done

---
In reply to: [132582648](https://github.com/Azure/azure-rest-api-specs/pull/1514#discussion_r132582648) [](ancestors = 132582648)

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (mtuchkov) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: Done

---
In reply to: [132580693](https://github.com/Azure/azure-rest-api-specs/pull/1514#discussion_r132580693) [](ancestors = 132580693)

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (mtuchkov) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: Done

---
In reply to: [132587892](https://github.com/Azure/azure-rest-api-specs/pull/1514#discussion_r132587892) [](ancestors = 132587892)

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (mtuchkov) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: Done

---
In reply to: [132583618](https://github.com/Azure/azure-rest-api-specs/pull/1514#discussion_r132583618) [](ancestors = 132583618)

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (mtuchkov) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: Done

---
In reply to: [132580819](https://github.com/Azure/azure-rest-api-specs/pull/1514#discussion_r132580819) [](ancestors = 132580819)

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (mtuchkov) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: No.

---
In reply to: [132580997](https://github.com/Azure/azure-rest-api-specs/pull/1514#discussion_r132580997) [](ancestors = 132580997)

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (mtuchkov) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: Done

---
In reply to: [132583662](https://github.com/Azure/azure-rest-api-specs/pull/1514#discussion_r132583662) [](ancestors = 132583662)

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (mtuchkov) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: Done

---
In reply to: [132584530](https://github.com/Azure/azure-rest-api-specs/pull/1514#discussion_r132584530) [](ancestors = 132584530)

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (mtuchkov) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: No

---
In reply to: [132581133](https://github.com/Azure/azure-rest-api-specs/pull/1514#discussion_r132581133) [](ancestors = 132581133)

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (mtuchkov) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: Done

---
In reply to: [132583721](https://github.com/Azure/azure-rest-api-specs/pull/1514#discussion_r132583721) [](ancestors = 132583721)

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (mtuchkov) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: No

---
In reply to: [132583788](https://github.com/Azure/azure-rest-api-specs/pull/1514#discussion_r132583788) [](ancestors = 132583788)

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (anuchandy) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: is there any properties of this model that server can set but client is not allowed to change/set (e.g. `created` seems a property that service set right?).. could you mark those properties as readonly?

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (anuchandy) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: could you mark relevant properties as `readonly` (if  any)?

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (mtuchkov) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: All of them are readonly

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (mtuchkov) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: Done

Comment (anuchandy) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: Could you fix the indentation?

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (anuchandy) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: Why there is no `id` field for this model `CertificateDescription`? It looks like it can have id with standard ARM format `/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Devices/IotHubs/{resourceName}/certificates/{certificateName}`


[1]. With id, consumer should be able to extract resource group name, parent iot hub name. This can be further used to perform update or delete operation on a certificate instance, because those actions required these fields.

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (anuchandy) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: [1]. Is this a paged collection? i.e. does service return all the certificates in a single response or page by page with a page containing link to fetch next page? see https://github.com/Azure/azure-rest-api-specs/blob/current/documentation/swagger-extensions.md#x-ms-pageable for more on paged collection. 

[2]. Is there any limit on number of certificates that can be created under an IoT hub resource?

**Comment by mtuchkov** [comment]
@anuchandy another PR was for previous version of API. it is not needed to change the old version as long as we have shipped a new version meanwhile.

Linter errors and warnings are not applicable to us. It requires to add fields like location and properties, but this is a child resource, we do not have these properties on child resource. 

other warnings about PUT and GET data contract, it asks to make them identical. But they cannot be the same, User uploads certificate and we return them only certificate properties, but not the cert itself.

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (mtuchkov) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: This is not a paged collection. The limit is 10.

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (mtuchkov) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: certificateName is the unique identified. the name is used for update and delete operations.

Comment (anuchandy) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: I was thinking from users point of view, how he can get parent information (resource group and IoTResource) from certificate resource so that he can use Delete & Update APIs easily. 

Otherwise consumer endup enumerating through all resource groups, IoTResources in RGs and certificates in IoTResource, in-order to retrieve the parent resource names required for Delete, Update API, right?

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (anuchandy) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: Thanks for confirming

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (mtuchkov) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: I guess I misunderstood you. Do you suggest to add the id field that is the full resource Id (i.e. `/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Devices/IotHubs/{resourceName}/certificates/{certificateName}`)?

Comment (anuchandy) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: yes, correct.

Comment (anuchandy) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: Update: Discussed with Mikhail offline, an issue has been created in the service side to fix this in the future releases.

Comment (ravbhatnagar) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: @mtuchkov - id, name and type properties MUST be present for /certificates resource. "Id" is needed for RBAC to work. This is a blocking issue. Please fix on the service side before we can onboard you.

**Review by ravbhatnagar — CHANGES_REQUESTED** [review]
@mtuchkov - Please ensure that the "certificates" resources complies with the basic shape that is required by ARM. Please add the id, name type properties and move all other properties inside the properties bag.

**Comment by mtuchkov** [comment]
@ravbhatnagar Please take a look

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (mtuchkov) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: Certificate operations are pass through calls (actions) on IotHub resource. These pass through calls will not modify IoTHub's resource etag. Do we still need to track it as a nested resource?

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (amarzavery) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: make sure indentation is 2 spaces everywhere

**Comment by salameer** [comment]
@ravbhatnagar 

I know we've chatted about this with the group are we good to go from your side?

**Comment by ravbhatnagar** [comment]
@salameer - as discussed over email, we don't want SDKs/docs to be generated for this. SO I would say, lets hold off on merging this.

**Comment by mtuchkov** [comment]
@ravbhatnagar what's left to merge this change?

File: specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json
Comment (mtuchkov) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: Done

Comment (mtuchkov) [inline specification/iothub/resource-manager/Microsoft.Devices/2017-07-01/iothub.json]: Done

**Comment by salameer** [comment]
SDK is OK with merging if Gaurav gives the Greenlight from ARM

**Comment by salameer** [comment]
@mtuchkov I know that you guys are waiting for the service to go live before the changes get merged in, what's the status here?

**Comment by veronicagg** [comment]
@mtuchkov any status updates?

**Comment by salameer** [comment]
I have email confirmation from Shankar that the service has been deployed, I think we're unlbokced on Merging :)

**Comment by anuchandy** [comment]
@mtuchkov seems fix is deployed.. could you update the certificateDescription to reflect the fix?

**Comment by anuchandy** [comment]
@mastermanu should I close this PR given that you have a new PR with same features?

**Comment by anuchandy** [comment]
This PR is not relevant any more, closing this.

**Pull request closed without merging**
